File: src/test/java/dev/kensa/matcher/Conditions.java
package dev.kensa.matcher;

import org.assertj.core.api.Condition;

import java.util.function.Function;
import java.util.function.Predicate;

public final class Conditions {

    public static <T, EXPECTED> Condition<T> equalTo(String description, EXPECTED expected, Predicate<T> predicate) {
        return new Condition<>(
                predicate,
                String.format("a %s of %%s", description),
                expected
        );
    }

    public static <T, EXPECTED> Condition<T> equalTo(EXPECTED expected, Predicate<T> predicate) {
        return new Condition<>(
                predicate,
                String.format("a %s of %%s", unCamelClassOf(expected)),
                expected
        );
    }

    public static <T, EXPECTED> Condition<T> equalTo(Function<T, EXPECTED> extractor, EXPECTED expected) {
        return equalTo(expected, Predicates.equalTo(extractor, expected));
    }

    public static <T, EXPECTED> Condition<T> equalTo(String description, Function<T, EXPECTED> extractor, EXPECTED expected) {
        return equalTo(description, expected, Predicates.equalTo(extractor, expected));
    }

    public static <T, EXPECTED> Condition<T> matches(Function<T, EXPECTED> extractor, Condition<EXPECTED> subCondition, EXPECTED expected) {
        return new Condition<>(
                t -> subCondition.matches(extractor.apply(t)),
                String.format("a %s matching %%s", unCamelClassOf(expected)),
                expected
        );
    }

    private static String unCamel(String camelCasedWords) {
        return String.join(" ", camelCasedWords.split("(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])"));
    }

    private static String unCamel(Class<?> c) {
        return unCamel(c.getSimpleName());
    }

    private static String unCamelClassOf(Object o) {
        return o == null ? "Null Instance" : unCamel(o.getClass().getSimpleName());
    }

    private Conditions() {
    }
}

File: src/test/java/dev/kensa/matcher/Conditions.kt
package dev.kensa.matcher

import org.assertj.core.api.Condition
import java.util.function.Predicate
import kotlin.reflect.KClass

object Conditions {
    fun <T, EXPECTED> equalTo(description: String, expected: EXPECTED, predicate: Predicate<T>): Condition<T> =
            Condition(predicate, "a $description of %%s", expected)

    fun <T, EXPECTED> equalTo(expected: EXPECTED, predicate: Predicate<T>): Condition<T> =
            Condition(predicate, "a ${unCamelClassOf(expected)} of %%s", expected)

    fun <T, EXPECTED> equalTo(extractor: (T) -> EXPECTED, expected: EXPECTED): Condition<T> =
            equalTo(expected, Predicates.equalTo(extractor, expected))

    @JvmStatic
    fun <T, EXPECTED> equalTo(description: String, extractor: (T) -> EXPECTED, expected: EXPECTED): Condition<T> =
            equalTo(description, expected, Predicates.equalTo(extractor, expected))

    fun <T, EXPECTED> matches(extractor: (T) -> EXPECTED, subCondition: Condition<EXPECTED>, expected: EXPECTED): Condition<T> =
            Condition(
                    Predicate { t: T -> subCondition.matches(extractor(t)) }, "a ${unCamelClassOf(expected)} matching %%s",
                    expected
            )

    private fun unCamel(camelCasedWords: String): String = camelCasedWords.split("(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])").joinToString(separator = " ")

    private fun unCamel(c: KClass<*>): String = unCamel(c.simpleName!!)

    private fun unCamelClassOf(o: Any?): String = if (o == null) "Null Instance" else unCamel(o::class.simpleName!!)
}